generated/nidmm/nidmm.proto

```diff
@@ -224,14 +224,14 @@ enum AcquisitionStatus {
 
 enum ApertureTime {
```

Comment:
I agree with using hashes on string values since there really isn't any good human-readable numeric representation of the string we could use. But for these enum values that fit nicely into 32-bit integer ranges, it's a shame to obfuscate them with hashed keys. 

I think there's a way we can release dmm and dcpower with the quick fix Zach was about to make ([ni/grpc-device PR 212](https://github.com/ni/grpc-device/pull/212/files)) but provide a backwards-compatible way to hash floats in the future for specific enums when their values can't be represented as 32-bit integers.

Here's my proposal:
- For string enums, always generate mappings using hashes as keys
- For a `ViReal64` enum that has `generate_mappings` set on it, generate hashes as keys
- For a `ViReal64` enum that does not have `generate_mappings` and no values have `map_key` on them, we would use the values as-is and perform a `static_cast` between `double` and `int` like Zach did in his PR.
- For a `ViReal64` enum that does not have `generate_mappings` but one or more of its values have `map_key` defined on it, we generate the map in the service handler with only those values tagged with a `map_key`. If a lookup in the map fails, we fall back to Zach's cast.
- For the `ViReal64` attribute enum, during the metadata mutation step, we would generate hashed keys for the enums tagged with `generate_mappings` and put those keys in `map_key`. This allows our attribute enum to be a mix of mapped and unmapped values. The mapped values would be somewhat rare since only a few of those floating-point enums don't work as ints.

This approach allows us to release with what we have now for the short term and buys us time to work out the best way to hash the float values into keys.

Reply:
I like this route. Seems like a good hybrid to be able to support existing proto enum values while providing backwards-compatibility.

### The below is probably out of scope and might be missing things but I wanted to get it out as I've been thinking about it the past few days

I've been mulling over this for a bit since that meeting we had and keep trying to think of an alternative to not have to have hashes in the proto file. The below might be out of scope, missing a concern the hashes address, and there might be holes in it that I'm not considering but it might be an alternative and thought I'd lay it out while it was on my mind:

I keep going back and thinking about how these types of concerns (unique constant to grpc Enum value mappings) could be captured in the metadata files in a simpler way somehow. One concern I have in general with this hashing going into the proto file is that over the wire smaller positive integers are more efficient on Enums than large numbers and negative numbers due to it using varint encoding ([see here in the proto3 docs for enums](https://developers.google.com/protocol-buffers/docs/proto3#enum)).  Also, any metadata mutations we make could be done separately and stored in the metadata files themselves and saved off so they become a starting point for each driver with the following form:

```
enums = {
    'RegularEnumForConstantGroup': {
        'value_type' : "std::string",
        'values': [
            {
                'name' : 'RegularEnumForConstantGroup_Unspecified'
                'grpc_value' : 0,
                'value' : <some default for enum value type>
            },
            {
                'name': 'RegularEnumForConstantGroup_ValueOne',
                'grpc_value': 1,
                'value': 'Some_Value_One',
            },
            {
                'name': 'RegularEnumForConstantGroup_ValueTwo',
                'grpc_value': 2,
                'value': 'Some_Value_Two'
            },
        ]
    },
    'VIInt32AttributeEnum': {
        'value_type' : "VIInt32",
        'values': [
            {
                'name' : 'VIInt32AttributeEnum_Unspecified'
                'grpc_value' : 0,
                'value' : <some default for enum value type>
            }
            {
                'name': 'VIInt32AttributeEnum_ValueOne',
                'grpc_value': 1,
                'value': -230
            }
        ]
    },
    'VIReal64AttributeEnum': {
        'value_type' : "VIReal64",
        'values': [
            {
                'name' : 'VIReal64AttributeEnum_Unspecified'
                'grpc_value' : 0,
                'value' : <some default for enum value type>
            }
            {
                'name': 'VIReal64AttributeEnum_ValueOne',
                'grpc_value': 1,
                'value': 1.23
            }
            {
                'name': 'VIReal64AttributeEnum_ValueTwo',
                'grpc_value': 2,
                'value': -123.45
            }
        ]
    },
}
```

Then, to make it simple in proto we always put `name` : `grpc_value`. In service handlers, we always have maps to translate `grpc_value` to `value` and back for every enum. This may be redundant for some int32 enums but it keeps it consistent in that the client is trying to pass over some representation of a constant and isn't concerned with passing the right value over the wire, just that it gets interpreted as expected on the server side.

Then later, when we get grpc_metadata generation in place name updates should be fine as old names will still translate to same grpc_value over the wire so I believe that should work. I suppose if a constant's actual value changes we'll want old callers trying to pass that constant to get the new value so that should be fine.
If there's new values, we make sure to take into account last 'grpc_value' or 'reserved' value in the previous "version" of a given enum in enums.py and increment for each new constant we're adding as an enum value to the metadata file. For removals, I think we treat it as the same as removals from a function parameter list: for backwards compatibility mark the removed values and possibly names as `reserved`.

Also, if the concern about backwards compatibility is instead that we want users using an old version of the proto to get the same value if they use old constants that don't exist anymore in the C API Header file I guess we could just never remove them from enums.py, just expand upon it so old values are always supported.

Reply:
@jonlugoatx I liked your approach of selectively hashing only those real64 constants which can't be represented using 32 bits. This also addresses Ryan's concern about "using larger int hash is costlier over wire compared to small numeric values". If we go that way, do we even need users having to tag enums and enum values with `generate-mappings` and `map_keys`? This is something our codegen script can detect and tag during mutation automatically. This way we don't have to update our metadata manually.

@reckenro the approach you have mentioned is basically maintaining state of last codegen w.r.t. enum mappings and using that for new enums going forward. Using `generate-mappings` for all enums (including int32 ones) will break backward compatibility though. We definitely can't use it with selective mapping Jonathan has mentioned above, but we can use it for specific type of enums. We can maintain generated cache under respective drivers' metadata folder (`enums.generated.cache`?) which contains dictionary of `actual_value: grpc_value`. New entries would get added with `max(grpc_value)+1` value. We probably don't even have to worry about cross-driver synchronization of this dictionary. We don't have to worry about names either, because they are not used for client-server communication. We probably shouldn't care about value changed/removed from header either, since that is essentially driver breaking the API and behavior of old client trying to communicate with incompatible new server (or other way around) would be undefined. Not removing entry from above cache would take care of us not reusing old obsoleted `grpc_value`.

Here's what I am thinking:
- Real64 enums - Use selective mapping approach Jonathan proposed (with long hash values for mapped values). Since we should have only few values (is this assumption correct?) that don't fit in 32 bit, we shouldn't have to worry about wire transfer performance. This gives us benefit of having to obfuscate only few values.
- String enums - Use cached index approach Ryan mentioned for all values.
- Int32/Int64 enums - No mapping
- No need to tag `generate_mappings` and `map_key` in metadata file (but we can still provide override if needed in future)

Downside of selective mapping:
1. Inconsistent values (few mapped/few unmapped) in proto file. Users could get confused by this.
2. If new client tries to pass a (new) mapped value to old server, server would treat it as raw value since it can't find it in dictionary resulting in undefined behavior.